File: Source/AdventOfCode2020/Problems/Problem3.cs
namespace AdventOfCode2020.Problems
{
	using System.Collections.Generic;
	using AdventOfCode2020.Utils.Extensions;

    /// <summary>
    /// Solution for <a href="https://adventofcode.com/2020/day/3">Day 3</a>.
    /// </summary>
    public class Problem3 : ProblemBase
    {
        public Problem3() : base(3) { }

        /// <inheritdoc />
        protected override object SolvePartOne()
        {
            return CountHits(ParseInputToMap(Input.WithoutEmptyLines()), 3, 1);
        }

        /// <inheritdoc />
        protected override object SolvePartTwo()
        {
            var map = ParseInputToMap(Input.WithoutEmptyLines());

            long total = 1;

            total *= CountHits(map, 1, 1);
            total *= CountHits(map, 3, 1);
            total *= CountHits(map, 5, 1);
            total *= CountHits(map, 7, 1);
            total *= CountHits(map, 1, 2);

            return total;
        }

        internal static bool[,] ParseInputToMap(IList<string> input)
        {
            var result = new bool[input[0].Length, input.Count];

            for (var y = 0; y < input.Count; y++)
            {
                for (var x = 0; x < input[y].Length; x++)
                {
                    result[x, y] = input[y][x] == '#';
                }
            }

            return result;
        }

        internal static int CountHits(bool[,] map, int xSpeed, int ySpeed)
        {
            var (x, y) = (0, 0);

            var trees = 0;

            while (y < map.GetLength(1))
            {
                if (map[x%map.GetLength(0), y])
                {
                    trees++;
                }

                x += xSpeed;
                y += ySpeed;
            }

            return trees;
        }
    }
}

File: java/src/main/com/adventofcode2020/Problems/Problem3.java
package com.adventofcode2020.Problems;

import java.util.List;

public class Problem3 extends ProblemBase {
    public Problem3() {
        super(3);
    }

    @Override
    public void solve() {
        var map = parseInputToMap(readInputAsStrings());

        System.out.println("3.1: " + solvePartOne(map));
        System.out.println("3.2: " + solvePartTwo(map));
    }

    public static Integer solvePartOne(boolean[][] map) {


        return 0;
    }

    public static Integer solvePartTwo(boolean[][] map) {


        return 0;
    }

    public static boolean[][] parseInputToMap(List<String> input)
    {
        var result = new boolean[input.get(0).length()][input.size()];

        for (var y = 0; y < input.size(); y++)
        {
            for (var x = 0; x < input.get(y).length(); x++)
            {
                result[x][y] = input.get(y).charAt(x) == '#';
            }
        }

        return result;
    }
}
